sdks/python/apache_beam/runners/worker/dataflow_element_execution_tracker.pyx

```diff
@@ -0,0 +1,330 @@
+    cdef unordered_map[ElementExecutionPtr, int64_t].iterator it = (
+      executions_per_element.begin())
+    cdef ElementExecutionPtr element_ptr = NULL
+    cdef int64_t attribution_time = 0
+    # Attribute processing time.
+    while it != executions_per_element.end():
+      element_ptr = dref(it).first
+      attribution_time = (sample_time / total_execution
+                         * executions_per_element[element_ptr])
+      self.execution_duration[element_ptr] += attribution_time
+      iref(it)
+
+  cdef void update_counter_cache(self, int64_t latest_snapshot,
+      unordered_map[CharPtr, vector[int64_t]]& counter_cache) nogil:
+    """Update counter_cache with values of done elements up to the given 
+    snapshot and prune done_journals.
+    """
+    cdef SnapshottedExecutionPtr snapshotted_ptr = NULL
+    cdef ElementExecutionPtr element_ptr = NULL
+    while self.shared_state.done_journal.journal.empty() is False:
+      snapshotted_ptr = self.shared_state.done_journal.journal.front()
+      if snapshotted_ptr.snapshot <= latest_snapshot:
+        self.shared_state.done_journal.journal.pop_front()
+        element_ptr = snapshotted_ptr.execution_ptr
+        # Unit of sampling time is nano secs, we update counter_cache with ms.
+        (counter_cache[element_ptr.operation_name]
+         .push_back(self.execution_duration[element_ptr]/1000000))
+      else:
+        break
+
+
+cdef class ExecutionJournalWriter(object):
+  """Writes journal entries on element processing state changes.
+
+  Attributes:
+    shared_state: Shared journals with ExecutionJournalReader.
+    execution_stack: Execution stack of processing elements. Elements are pushed
+    onto the stack when scoped_process_state.__enter__ and popped on
+    scoped_process_state.__exit__. This stack mirrors the actual runtime stack
+    and contains the step + element context in order to attribute sampled
+    execution time.
+
+  Args:
+    shared_state: Initialized in DataflowElementExecutionTracker.
+  """
+  def __init__(self, ReaderWriterState shared_state):
+    self.shared_state = shared_state
+    self.execution_stack = deque[ElementExecutionPtr]()
+    self.add_execution(NULL)
+
+  cdef void add_execution(self, ElementExecutionPtr execution_ptr):
+    """Push current ElementExecution onto top of the execution_stack and add
+    new entry into journal.
+    """
+    cdef int64_t next_snapshot = self.shared_state.latest_snapshot + 1
+    self.execution_stack.push_back(execution_ptr)
+    self.shared_state.execution_journal.add_journal(execution_ptr,
+                                                    next_snapshot)
+    self.shared_state.latest_snapshot = next_snapshot
+
+  cdef void start_processing(self, char* operation_name):
+    """Create and journal a new ElementExecution to track a processing element.
+    """
+    cdef ElementExecutionPtr execution_ptr = (
+      <ElementExecutionPtr> malloc(sizeof(ElementExecution)))
```

Comment:
I'm not sure what a ```pool``` means here. The reasons why I alloc memory dynamically are: 
1. It's faster to move pointer here and there rather than copying item directly.
2. At this point, we may need memory alloced in heap rather than stack.

Reply:
I was thinking that we can allocate a larger object in the heap first, and use indexes into that rather than doing bunch of small allocations.

> It's faster to move pointer here and there rather than copying item directly.

Note that, the item (`ElementExecution`) is `char* operation_name` is just a struct wrapping around a `char*`. We might as well pass pointers to the operation_name instead of allocating another object. Do we plan on adding more items into `ElementExecution`?

Reply:
For ```ElementExecution```:
1. We don't have a plan adding more attr into this struct.
2. The reason why we need to keep a ```ElementExecution``` is, we want to track processing time per element. So the ```ElementExecutionPtr``` actually is a object that marks an element with operaion_name.
3. If we can do the pre-allocating things, that would be better. I'm concerning, we cannot know the size of pool at very beginning, which means we may need resize the pool at some point.

Reply:
I think you can still use `ElementExecutionPtr` by typedefing it to `char*`, would it work?

Let's not do 3 in this PR. One can do malloc's to extend the size of this large object when needed. This will happen much less frequently than individually allocation memory for each element.